File: python/gripsack/tree.py
"""Tree entries — directory-shaped config deploys via eval-time
expansion (0006 critique; the IR stays per-file).

>>> from gripsack import module, tree  # doctest: +SKIP
>>> module("zed", config={**tree("configs/zed", "~/.config/zed")})  # doctest: +SKIP

Adding or removing files in the directory is picked up at the next
eval; files dropped from the tree are pruned at apply (0008 §3).
"""

from __future__ import annotations

from pathlib import Path

from .entries import Dest, Ownership


def tree(src: str, to: str, mode: Ownership = Ownership.TRACKED_COPY) -> dict[str, Dest]:
    """Expand a directory into per-file config entries: every file under
    `src` maps to its mirror under `to`, with `mode`.

    Merge the result into a module's config: `config={**tree(...)}`.
    """
    root = Path(src)
    entries: dict[str, Dest] = {}
    for path in sorted(root.rglob("*")):
        if path.is_file() and not path.is_symlink():
            rel = path.relative_to(root).as_posix()
            entries[f"{src}/{rel}"] = Dest(f"{to}/{rel}", mode)
    return entries

File: typescript/src/tree.ts
/** Tree entries — directory-shaped config deploys via eval-time
 *  expansion (the IR stays per-file). Adding or removing files in the
 *  directory is picked up at the next eval; files dropped from the
 *  tree are pruned at apply.
 *
 *  ```ts
 *  module("zed", { config: { ...tree("configs/zed", "~/.config/zed") } });
 *  ```
 */

import { readdirSync, statSync } from "node:fs";
import { join, relative } from "node:path";
import type { Dest, Ownership } from "./entries.ts";

export function tree(
  src: string,
  to: string,
  mode: Ownership = "tracked_copy",
): Record<string, Dest> {
  const entries: Record<string, Dest> = {};
  const walk = (dir: string): void => {
    for (const name of readdirSync(dir).sort()) {
      const path = join(dir, name);
      if (statSync(path).isDirectory()) {
        walk(path);
      } else if (statSync(path).isFile()) {
        const rel = relative(src, path).split("\\").join("/");
        entries[`${src}/${rel}`] = { to: `${to}/${rel}`, mode };
      }
    }
  };
  walk(src);
  return entries;
}
